### PR TITLE
[runtime]add nd.reshape

### PR DESCRIPTION
**Pull request by jc-bytedance**

### CODE REVIEW

File: python/matx/ir/op.py [hunk 1]
```diff
@@ -1986,6 +1986,16 @@ def object_contiguous(span, container_expr, *args, **kwargs):
     return hlo_call_intrin(ret_type, func_name, span, container_expr, *args, **kwargs)
 
 
+def object_reshape(span, container_expr, *args, **kwargs):
+    ret_type = _type.ObjectType()
+    func_name = _builtin_func_name(container_expr, "reshape")
+    if _type_rel.is_type_of(container_expr, _type.NDArrayType):
+        ty = container_expr.py_type_name()
+        assert len(kwargs) == 0, f"{ty}.transpose() takes no keyword arguments"
```
Comment (cloud-mxd) [inline python/matx/ir/op.py:1994]: typo "transpose"

File: src/c_api/ndarray_c_api.cc [hunk 2]
```diff
@@ -52,6 +52,36 @@ MATXSCRIPT_REGISTER_GLOBAL("runtime.NDArrayContiguous").set_body([](PyArgs args)
...
+  std::vector<int64_t> shape;
+  switch (args[1].type_code()) {
+    case TypeIndex::kRuntimeList: {
+       auto it = args[1].AsObjectRefNoCheck<List>();
+       for(auto &e:it){
+          shape.push_back(e.As<int64_t>());
+       }
+    }break;
+    case TypeIndex::kRuntimeFTList: {
+       auto it = args[1].AsObjectRefNoCheck<FTObjectBase>();
```
Comment (cloud-mxd) [inline src/c_api/ndarray_c_api.cc:66]: Maybe just need to support FTList[int]: As<FTList<int64>>()

File: src/runtime/container/ndarray.cc [hunk 3]
```diff
@@ -566,6 +566,35 @@ struct NDArray::Internal {
   }
 };
 
+NDArray NDArray::Reshape(std::vector<int64_t> shape) const {
```
Comment (cloud-mxd) [inline src/runtime/container/ndarray.cc:569]: This interface needs to support three types: List/Tuple/Any. 
STL types do not necessarily support.

File: test/script/test_ndarray.py [hunk 4]
```diff
@@ -801,6 +801,32 @@ def generic_contiguous(x: Any) -> Any:
         self.assertEqual(z0[0][0], 1)
         self.assertEqual(z1[0][0], 1)
 
+    def test_nd_reshape(self):
+        def generic_reshape(x: Any, newshape: Any) -> Any:
+            return x.reshape(newshape)
+
+        def specific_shapre(x: matx.NDArray, newshape: tuple) -> matx.NDArray:
+            return x.reshape(newshape)
```
Comment (cloud-mxd) [inline test/script/test_ndarray.py:809]: Need to add a test: NDArray.reshape([...])

File: python/matx/ir/op.py (hunk 1 above)
Comment (jc-bytedance) [reply python/matx/ir/op.py]: fixed

File: src/c_api/ndarray_c_api.cc (hunk 2 above)
Comment (jc-bytedance) [reply src/c_api/ndarray_c_api.cc]: fixed

File: test/script/test_ndarray.py (hunk 4 above)
Comment (jc-bytedance) [reply test/script/test_ndarray.py:812]: added new test

File: src/runtime/container/ndarray.cc (hunk 3 above)
Comment (jc-bytedance) [reply src/runtime/container/ndarray.cc]: added List/Tuple/Any/FTList

File: src/c_api/ndarray_c_api.cc [hunk 5]
```diff
@@ -52,6 +52,12 @@ MATXSCRIPT_REGISTER_GLOBAL("runtime.NDArrayContiguous").set_body([](PyArgs args)
   return data.Contiguous();
 });
 
+MATXSCRIPT_REGISTER_GLOBAL("runtime.NDArrayReshape").set_body([](PyArgs args) -> RTValue {
+  NDArray data = args[0].As<NDArray>();
+  Any newshape = args[1];
```
Comment (cloud-mxd) [inline src/c_api/ndarray_c_api.cc:57]: This assignment is unnecessary

File: src/runtime/container/ndarray.cc [hunk 6]
```diff
@@ -566,6 +566,81 @@ struct NDArray::Internal {
   }
 };
 
+NDArray NDArray::Reshape(std::vector<int64_t>& newshape) const {
```
Comment (cloud-mxd) [inline src/runtime/container/ndarray.cc:569]: remove '&', and use std::move instead

**Review by cloud-mxd — APPROVED** [review]

**Pull request merged**